### PR TITLE
feat:Apis for user Authentication

### PR DESCRIPTION
**Pull request by jaimin45**
Added Apis user authentication.

### CODE REVIEW

File: swagger.yaml [hunk 1]
```diff
@@ -124,6 +124,136 @@ paths:
           description: "Internal Server Error"
           schema:
             $ref: "#/definitions/500Response"
+  /user:
```
Comment (kapilbarad) [inline swagger.yaml]: Should be `/users`

File: swagger.yaml [hunk 2]
```diff
@@ -124,6 +124,136 @@ paths:
...
+    post:
+      tags:
+        - User
+      summary: Register user into system
+      parameters:
+        - name: user
+          in: body
+          description: Create user
+          schema:
+            $ref: "#/definitions/user"
```
Comment (kapilbarad) [inline swagger.yaml]: We should define definitions for Create or Update body separately it should be not just `User` which we will use in responses.

File: swagger.yaml [hunk 3]
```diff
@@ -124,6 +124,136 @@ paths:
...
+      responses:
+        "201":
+          description: OK
```
Comment (kapilbarad) [inline swagger.yaml]: Description should be `Created`, as it's 201 status code

File: swagger.yaml [hunk 4]
```diff
@@ -124,6 +124,136 @@ paths:
...
+        - name: email
+          in: path
+          required: true
+          description: id of the User
+          type: string
+      responses:
+        "200":
+          description: otp send on email
+          schema:
+            $ref: "#/definitions/user"
```
Comment (kapilbarad) [inline swagger.yaml]: Wrong schema.

File: swagger.yaml [hunk 5]
```diff
@@ -124,6 +124,136 @@ paths:
...
+        - name: User
+          in: body
+          description: User password reset though id using ott
+          schema:
+            $ref: "#/definitions/userResetPassword"
+      responses:
+        "204":
+          description: Password is updated
+          schema:
+            $ref: "#/definitions/user"
```
Comment (kapilbarad) [inline swagger.yaml]: No need of schema.

File: swagger.yaml [hunk 6]
```diff
@@ -124,6 +124,136 @@ paths:
...
+      responses:
+        "201":
+          description: OK
+          schema:
+            $ref: "#/definitions/user"
+        "500":
+          description: "Internal Server Error"
+          schema:
+            $ref: "#/definitions/500Response"
+  /user/reset-password:
```
Comment (kapilbarad) [inline swagger.yaml]: This should be actually in Forgot Password API.
Reset Password API should accept Current Password and New Password. if the current Password is valid then update new password.

File: swagger.yaml [hunk 7]
```diff
@@ -124,6 +124,136 @@ paths:
...
+    get:
```
Comment (kapilbarad) [inline swagger.yaml]: Should be PATCH.

File: swagger.yaml [hunk 8]
```diff
@@ -124,6 +124,136 @@ paths:
...
+            $ref: "#/definitions/400Response"
+        "404":
+          description: "User not found"
+          schema:
+            $ref: "#/definitions/404Response"
+        "500":
+          description: "Internal Server Error"
+          schema:
+            $ref: "#/definitions/500Response"
+    post:
```
Comment (kapilbarad) [inline swagger.yaml:201]: This should be under `/user/forgot-password`

File: swagger.yaml [hunk 9]
```diff
@@ -124,6 +124,136 @@ paths:
...
+        "404":
+          description: "User not found"
+          schema:
+            $ref: "#/definitions/404Response"
+        "500":
+          description: "Internal Server Error"
+          schema:
+            $ref: "#/definitions/500Response"
+
+  /user/login:
```
Comment (kapilbarad) [inline swagger.yaml]: Path should be `/oauth`

File: swagger.yaml [hunk 10]
```diff
@@ -124,6 +124,136 @@ paths:
...
+    get:
```
Comment (kapilbarad) [inline swagger.yaml]: Method should be POST as we're creating a new token.

File: swagger.yaml [hunk 11]
```diff
@@ -124,6 +124,136 @@ paths:
...
+      tags:
+        - User
+      summary: Login user into system
+      parameters:
+        - name: "username"
```
Comment (kapilbarad) [inline swagger.yaml]: Should be email.

File: swagger.yaml [hunk 12]
```diff
@@ -124,6 +124,136 @@ paths:
...
+          description: "The user name for login"
+          required: true
+          type: "string"
+        - name: "password"
+          in: "query"
+          description: "The password for login in clear text"
+          required: true
+          type: "string"
+      responses:
+        "200":
```
Comment (kapilbarad) [inline swagger.yaml]: Should be 201

File: swagger.yaml [hunk 13]
```diff
@@ -124,6 +124,136 @@ paths:
...
+          description: OK
+          schema:
+            $ref: "#/definitions/userLogin"
```
Comment (kapilbarad) [inline swagger.yaml]: Definition Name should be OauthToken

File: swagger.yaml [hunk 14]
```diff
@@ -124,6 +124,136 @@ paths:
...
+        "404":
+          description: "Tutorial not found"
```
Comment (kapilbarad) [inline swagger.yaml]: Wrong Description

**Review by kapilbarad — CHANGES_REQUESTED** [review]
Please correct below APIs:
POST `/oauth/token`: API to create JWT Token
POST `/oauth/forgot-password`: API to create an OTT and send to email
PATCH `/oauth/forgot-password`: API to update password using OTT
PATCH `/oauth/reset-password`: API to update password using old password

CC: @ashishkava @JainishaParmar

File: swagger.yaml [hunk 15]
```diff
@@ -14,6 +14,11 @@ consumes:
   - application/json
 produces:
   - application/json
+securityDefinitions:  
+  authentication:  
+    type: apiKey
```
Comment (kapilbarad) [inline swagger.yaml]: We should use JWT token in header - Authorization.

File: swagger.yaml [hunk 16]
```diff
@@ -107,6 +115,8 @@ paths:
           description: "Internal Server Error"
           schema:
             $ref: "#/definitions/500Response"
+      security:
```
Comment (kapilbarad) [inline swagger.yaml]: What about authentication in PATCH and  POST /tutorial?

File: swagger.yaml [hunk 17]
```diff
@@ -124,6 +134,162 @@ paths:
...
+    post:
+      tags:
+        - User
+      summary: Register user into system
+      parameters:
+        - name: createUser
+          in: body
+          description: Create user
+          schema:
+            $ref: "#/definitions/userCreated"
```
Comment (kapilbarad) [inline swagger.yaml]: Why the name is past sentence `userCreated`, This is request body not the response? Please refer to POST /tutorials API and follow same best practices.

File: swagger.yaml [hunk 18]
```diff
@@ -124,6 +134,162 @@ paths:
...
+      responses:
+        "201":
+          description: Created
+          schema:
+            $ref: "#/definitions/userCreated"
```
Comment (kapilbarad) [inline swagger.yaml]: Please follow definition name practices similar to /tutorial APIs

Comment (kapilbarad) [reply swagger.yaml]: Apply the same in other newly added APIs

File: swagger.yaml [hunk 19]
```diff
@@ -124,6 +134,162 @@ paths:
...
+        "500":
+          description: "Internal Server Error"
+          schema:
+            $ref: "#/definitions/500Response"
+
+    patch:
+      tags:
+        - User
+      summary: user updated into system
```
Comment (kapilbarad) [inline swagger.yaml]: This should be the the description of the API, The action which we're performing not the response.
```
Update user
```

File: swagger.yaml [hunk 20]
```diff
@@ -124,6 +134,162 @@ paths:
...
+          in: body
+          description: Create user
+          schema:
+            $ref: "#/definitions/userCreated"
+      responses:
+        "201":
+          description: Created
+          schema:
+            $ref: "#/definitions/userCreated"
+        "500":
```
Comment (kapilbarad) [inline swagger.yaml:156]: Missing 400 Bad Request response.

File: swagger.yaml [hunk 21]
```diff
@@ -124,6 +134,162 @@ paths:
...
+          in: body
+          required: true
+          description: Update user
+          schema:
+            $ref: "#/definitions/userUpdated"
+      responses:
+        "204":
+          description: user is updated
+        "400":
+          description: "Invalid user token"
```
Comment (kapilbarad) [inline swagger.yaml]: Wrong Description.

**Review by kapilbarad — CHANGES_REQUESTED** [review]
Please go through all the APIS again and correct the feedbacks.

**Review by JainishaParmar — APPROVED** [review]

**Review by ghost — APPROVED** [review]

File: swagger.yaml [hunk 22]
```diff
@@ -14,6 +14,11 @@ consumes:
   - application/json
 produces:
   - application/json
+securityDefinitions:
+  JWT:
+    type: apiKey
+    in: header
+    name: access_token
```
Comment (kapilbarad) [inline swagger.yaml]: USE `Authorization` header name.

File: swagger.yaml [hunk 23]
```diff
@@ -124,6 +129,166 @@ paths:
...
+    post:
+      tags:
+        - User
+      summary: Register user into system
+      parameters:
+        - name: createUser
+          in: body
+          description: Create user
+          schema:
+            $ref: "#/definitions/userCreateBody"
```
Comment (kapilbarad) [inline swagger.yaml]: All Definition's name should start with upper case: 'UserCreateBody'.

CC: @JainishaParmar @ashishkava

File: swagger.yaml [hunk 24]
```diff
@@ -124,6 +129,166 @@ paths:
...
+      responses:
+        "201":
+          description: created user
```
Comment (kapilbarad) [inline swagger.yaml:145]: Should be `User created`

File: swagger.yaml [hunk 25]
```diff
@@ -124,6 +129,166 @@ paths:
...
+          schema:
+            $ref: "#/definitions/user"
+        "400":
+          description: "Invalid user details"
```
Comment (kapilbarad) [inline swagger.yaml]: Should be Bad Request

File: swagger.yaml [hunk 26]
```diff
@@ -124,6 +129,166 @@ paths:
...
+      tags:
+        - User
+      summary: user updated into system
+      parameters:
+        - name: updateUser
+          in: body
+          required: true
+          description: Update user
+          schema:
+            $ref: "#/definitions/userUpdateBody"
```
Comment (kapilbarad) [inline swagger.yaml]: `UserUpdateBody`

**Review by kapilbarad — CHANGES_REQUESTED** [review]

**Review by JainishaParmar — APPROVED** [review]

**Review by kapilbarad — APPROVED** [review]

**Comment by jaimin45** [comment]
Feat : Apis for user Authentication

**Pull request merged**